File: shield_gh/crypto/pqc_lkh.cc
// ============================================================
// IMPLEMENTS: Eq. 3.34 — Kj = {(pku, sku) : u ∈ path(vj→root)}
//             Eq. 3.35 — (Kgrp, croot) = Kyber.Enc(pkroot, m)
//             Eq. 3.36 — (K^new_u, cu) = Kyber.Enc(pk^sib_u, k^new_u)
//                         ∀u ∈ path(ℓi → root)
// FIGURE 3.11: PQC-LKH Binary Tree for Post-Quantum Group Re-Keying
// ============================================================
#include "pqc_lkh.h"
#include <stdexcept>

// Build a complete binary tree for N vehicles.
// Leaf nodes correspond to individual vehicles.
// Internal nodes hold KEMs used for subgroup key encapsulation.
void PQCLogicalKeyHierarchy::Build(uint32_t N) {
    m_N = N;
    // Number of nodes in a complete binary tree with N leaves = 2N - 1
    uint32_t total_nodes = 2 * N - 1;
    m_tree.resize(total_nodes);

    // Generate key pairs for every node in the tree
    for (uint32_t i = 0; i < total_nodes; i++) {
        m_tree[i].id       = i;
        m_tree[i].key_pair = m_kyber.GenerateKeyPair();

        // Wire children and parent pointers (1-indexed complete binary tree)
        if (i > 0) {
            m_tree[i].parent = (i - 1) / 2;
        }
        uint32_t left  = 2 * i + 1;
        uint32_t right = 2 * i + 2;
        if (left  < total_nodes) m_tree[i].left_child  = left;
        if (right < total_nodes) m_tree[i].right_child = right;
    }

    // Assign vehicle IDs to leaf nodes (nodes N-1 ... 2N-2)
    for (uint32_t leaf = 0; leaf < N; leaf++) {
        uint32_t node_idx = (N - 1) + leaf;
        m_tree[node_idx].leaf_vehicle = leaf;
    }
}

// Find the tree node index for a given vehicle id (leaf lookup)
uint32_t PQCLogicalKeyHierarchy::GetLeafIndex(uint32_t vehicle_id) const {
    for (uint32_t i = 0; i < m_tree.size(); i++) {
        if (m_tree[i].leaf_vehicle == vehicle_id) return i;
    }
    throw std::runtime_error("Vehicle not found in LKH tree");
}

// Get the sibling of a given node index
uint32_t PQCLogicalKeyHierarchy::GetSibling(uint32_t node_idx) const {
    if (node_idx == 0) return 0;  // root has no sibling
    int32_t parent = m_tree[node_idx].parent;
    uint32_t left  = m_tree[parent].left_child;
    uint32_t right = m_tree[parent].right_child;
    return (node_idx == (uint32_t)left) ? right : left;
}

// ── Eq. 3.34 ─────────────────────────────────────────────────────────────────
// Kj = {(pku, sku) : u ∈ path(vj→root)}
// Returns node indices along the path from vehicle leaf to root
std::vector<uint32_t> PQCLogicalKeyHierarchy::GetPathToRoot(uint32_t vehicle_id) const {
    std::vector<uint32_t> path;
    uint32_t current = GetLeafIndex(vehicle_id);
    while (true) {
        path.push_back(current);
        if (m_tree[current].parent < 0) break;
        current = m_tree[current].parent;
    }
    return path;  // leaf → root order
}

// ── Eq. 3.35 ─────────────────────────────────────────────────────────────────
// (Kgrp, croot) = Kyber.Enc(pkroot, m),  m ←$ {0,1}^256
// Encapsulates the group session key under the root public key
std::pair<std::vector<uint8_t>, KyberCiphertext>
PQCLogicalKeyHierarchy::EncapsulateGroupKey() {
    // Root is always node 0
    return m_kyber.Encapsulate(m_tree[0].key_pair.pk);
}

// ── Eq. 3.36 ─────────────────────────────────────────────────────────────────
// When vehicle vi (leaf ℓi) is isolated:
// For each u ∈ path(ℓi → root): regenerate ku, encapsulate under sibling pk
// Cost: ⌈log2 N⌉ Kyber.Enc operations (vs O(N) unicast)
std::vector<KyberCiphertext>
PQCLogicalKeyHierarchy::IsolateAndRekey(uint32_t isolated_vehicle_id) {
    std::vector<KyberCiphertext> new_ciphertexts;

    std::vector<uint32_t> path = GetPathToRoot(isolated_vehicle_id);

    // Walk path from leaf to root, refreshing keys at each node u
    // Encrypt new key under the sibling's public key (sibling can then update)
    for (uint32_t node_idx : path) {
        // Regenerate key pair for this node (Eq. 3.36: K^new_u)
        m_tree[node_idx].key_pair = m_kyber.GenerateKeyPair();

        // Encapsulate new key under sibling's current public key
        uint32_t sib = GetSibling(node_idx);
        if (sib != node_idx) {
            auto [shared_secret, ct] = m_kyber.Encapsulate(m_tree[sib].key_pair.pk);
            new_ciphertexts.push_back(ct);
        }
    }

    return new_ciphertexts;  // |path| = ⌈log2 N⌉ ciphertexts
}

File: shield_gh_crypto/pqc_lkh.py
"""
SHIELD-GH Task 05 — PQC-LKH: Post-Quantum Logical Key Hierarchy
===============================================================
Report section 3.6.9 / sec:pqc_lkh ("PQC-LKH: Post-Quantum Logical Key
Hierarchy for Group Re-Keying"), Figure 3.11.  This is the "key sharing
(logical key hierarchy)" mechanism.

Equations implemented (with GENUINE Kyber KEM from pqc_primitives):
    K_j = {(pk_u, sk_u) : u in path(v_j -> root), leaf_u != VACANT}  # eq:lkh_keys (3.34)
    (K_grp, c_root) = Kyber.Enc(pk_root, m), m<-${0,1}^256           # eq:lkh_grp  (3.35)
    (K_u^new, c_u)  = Kyber.Enc(pk_u^sib, k_u^new),                  # eq:lkh_rekey(3.36)
                       for all u in path(leaf_i -> root)
Complexity: isolation / join re-keys ONLY the leaf->root path
    => exactly ceil(log2 N) Kyber operations, vs O(N-1) naive unicast.

Membership model (report "lazy policy"):
    Lazy Departure: an isolated/departed vehicle's leaf is marked VACANT, keeps
                    no private key, does not participate in future re-keying.
    Lazy Join:      a new vehicle takes the lowest-indexed VACANT leaf; its
                    root path is refreshed (ceil(log2 N) Kyber ops).

Security property demonstrated in tests & evidence:
    After IsolateAndRekey(v_i), EVERY remaining vehicle can decapsulate its way
    to the NEW group key K_grp', while the isolated v_i (whose held sk are all
    stale for refreshed path nodes) CANNOT -> cryptographic exclusion without
    unicast contact.

Complete binary tree layout (array form, 0-indexed):
    node 0 = root; children of i are 2i+1, 2i+2; parent of i is (i-1)//2.
    For N leaves we build a tree whose leaf layer has 2^ceil(log2 N) slots so a
    clean binary hierarchy exists; unused/departed slots are VACANT.
"""

from __future__ import annotations

import math
import secrets
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from pqc_primitives import KyberKEM, KyberKeyPair

VACANT = None


@dataclass
class LKHNode:
    idx: int
    keypair: Optional[KyberKeyPair] = None     # None => vacant leaf (no key)
    is_leaf: bool = False
    leaf_vehicle: Optional[int] = None         # vehicle id if occupied leaf


class PQCLogicalKeyHierarchy:
    def __init__(self, level: int = 768):
        """level = 512 (lightweight mode) or 768 (full mode) per report."""
        self.kem = KyberKEM(level)
        self.level = level
        self.nodes: List[LKHNode] = []
        self.depth = 0
        self.leaf_start = 0
        self.n_leaf_slots = 0
        self.group_key: Optional[bytes] = None
        # count of Kyber.Enc/Dec ops, to demonstrate O(log N) claim
        self.kyber_ops = 0

    # ------------------------------------------------------------------ #
    #  Build (Figure 3.11)                                               #
    # ------------------------------------------------------------------ #
    def build(self, vehicle_ids: List[int]) -> None:
        n = len(vehicle_ids)
        self.depth = max(1, math.ceil(math.log2(max(2, n))))
        self.n_leaf_slots = 2 ** self.depth
        total = 2 * self.n_leaf_slots - 1
        self.leaf_start = self.n_leaf_slots - 1
        self.nodes = [LKHNode(idx=i) for i in range(total)]

        # every internal node + occupied leaf gets a fresh Kyber keypair
        for i in range(total):
            is_leaf = i >= self.leaf_start
            self.nodes[i].is_leaf = is_leaf

        for slot, vid in enumerate(vehicle_ids):
            node = self.nodes[self.leaf_start + slot]
            node.leaf_vehicle = vid
            node.keypair = self.kem.generate_keypair()

        # internal nodes always carry keys
        for i in range(self.leaf_start):
            self.nodes[i].keypair = self.kem.generate_keypair()

    # ------------------------------------------------------------------ #
    #  Tree navigation                                                   #
    # ------------------------------------------------------------------ #
    @staticmethod
    def parent(i: int) -> int:
        return (i - 1) // 2 if i > 0 else -1

    def sibling(self, i: int) -> int:
        if i == 0:
            return -1
        p = self.parent(i)
        left, right = 2 * p + 1, 2 * p + 2
        return right if i == left else left

    def leaf_index_of(self, vehicle_id: int) -> int:
        for i in range(self.leaf_start, len(self.nodes)):
            if self.nodes[i].leaf_vehicle == vehicle_id:
                return i
        raise KeyError(f"vehicle {vehicle_id} not in tree")

    # ---- Eq. eq:lkh_keys (3.34):  K_j = {(pk_u,sk_u): u in path, !VACANT} --
    def path_to_root(self, vehicle_id: int) -> List[int]:
        """Node indices from the vehicle's leaf up to the root (Eq. 3.34)."""
        i = self.leaf_index_of(vehicle_id)
        path = []
        while i != -1:
            path.append(i)
            i = self.parent(i)
        return path

    def key_set(self, vehicle_id: int) -> List[int]:
        """K_j: the non-vacant nodes on v_j's root path whose sk it holds."""
        return [u for u in self.path_to_root(vehicle_id)
                if self.nodes[u].keypair is not None]

    # ------------------------------------------------------------------ #
    #  Eq. eq:lkh_grp (3.35): (K_grp, c_root) = Kyber.Enc(pk_root, m)     #
    # ------------------------------------------------------------------ #
    def encapsulate_group_key(self) -> Tuple[bytes, bytes]:
        root = self.nodes[0].keypair
        K_grp, c_root = self.kem.encapsulate(root.pk)
        self.kyber_ops += 1
        self.group_key = K_grp
        return K_grp, c_root

    def derive_group_key(self, vehicle_id: int, c_root: bytes) -> bytes:
        """Any vehicle holding sk_root (all do, before isolation) recovers K_grp."""
        root = self.nodes[0].keypair
        self.kyber_ops += 1
        return self.kem.decapsulate(root.sk, c_root)

    # ------------------------------------------------------------------ #
    #  Eq. eq:lkh_rekey (3.36): refresh ONLY path(leaf_i -> root)         #
    #  Lazy Departure: mark leaf VACANT, then re-key its ancestors.       #
    # ------------------------------------------------------------------ #
    def isolate_and_rekey(self, isolated_vehicle_id: int) -> Dict:
        """Isolate v_i and refresh exactly the leaf->root path.

        Returns a transcript dict:
          refreshed_nodes : list of node idx whose keypair was regenerated
          broadcasts      : list of (node_idx, sibling_idx, ciphertext c_u)
                            i.e. new key k_u^new encapsulated under sibling pk
          kyber_ops       : Kyber operations used (== ceil(log2 N))
        """
        path = self.path_to_root(isolated_vehicle_id)
        leaf_idx = path[0]

        # --- Lazy Departure: leaf becomes VACANT, loses its key material -----
        self.nodes[leaf_idx].keypair = VACANT
        self.nodes[leaf_idx].leaf_vehicle = None

        refreshed = []
        broadcasts = []
        ops_before = self.kyber_ops

        # walk ancestors leaf..root; refresh each and re-encapsulate under the
        # sibling subtree's CURRENT public key so only the non-isolated side can
        # obtain the new node key (Eq. 3.36).
        ancestors = path[1:]  # exclude the now-vacant leaf itself
        for u in ancestors:
            new_kp = self.kem.generate_keypair()          # k_u^new
            sib = self.sibling(u)
            if sib != -1 and self.nodes[sib].keypair is not None:
                # non-root level: encapsulate new node key under sibling pk
                # (Eq. 3.36) so only the unaffected subtree can obtain it.
                k_u_new, c_u = self.kem.encapsulate(self.nodes[sib].keypair.pk)
                self.kyber_ops += 1
                broadcasts.append((u, sib, c_u))
            else:
                # root level: no sibling. One Kyber.Enc distributes the fresh
                # group key K_grp' under the refreshed root pk (Eq. 3.35 re-run),
                # keeping the total at exactly ceil(log2 N) operations.
                self.nodes[u].keypair = new_kp
                K_grp_new, c_root = self.kem.encapsulate(new_kp.pk)
                self.kyber_ops += 1
                self.group_key = K_grp_new
                broadcasts.append((u, -1, c_root))
                refreshed.append(u)
                continue
            self.nodes[u].keypair = new_kp
            refreshed.append(u)

        ops = self.kyber_ops - ops_before
        return {
            "isolated": isolated_vehicle_id,
            "path": path,
            "refreshed_nodes": refreshed,
            "broadcasts": broadcasts,
            "kyber_ops": ops,
            "expected_log2N": self.depth,
        }

    # ---- Lazy Join -------------------------------------------------------
    def join(self, new_vehicle_id: int) -> Dict:
        """Assign new vehicle to lowest-indexed VACANT leaf; refresh its path."""
        slot = None
        for i in range(self.leaf_start, len(self.nodes)):
            if self.nodes[i].keypair is None:
                slot = i
                break
        if slot is None:
            raise RuntimeError("no VACANT leaf; tree full (report: depth grows)")
        self.nodes[slot].keypair = self.kem.generate_keypair()
        self.nodes[slot].leaf_vehicle = new_vehicle_id
        # refresh ancestors so the new member joins the hierarchy
        i = self.parent(slot)
        refreshed = []
        while i != -1:
            self.nodes[i].keypair = self.kem.generate_keypair()
            refreshed.append(i)
            i = self.parent(i)
        return {"joined": new_vehicle_id, "leaf_idx": slot,
                "refreshed_nodes": refreshed, "kyber_ops_est": self.depth}

    # ------------------------------------------------------------------ #
    #  Efficiency comparison (Figure 3.11 table)                         #
    # ------------------------------------------------------------------ #
    def rekey_cost(self) -> Dict:
        n = sum(1 for i in range(self.leaf_start, len(self.nodes))
                if self.nodes[i].keypair is not None)
        return {
            "N_active": n,
            "naive_unicast_ops": max(0, n - 1),   # O(N)
            "pqc_lkh_ops": self.depth,            # O(log N)
            "speedup": (max(1, n - 1) / max(1, self.depth)),
        }

    def active_vehicles(self) -> List[int]:
        return [self.nodes[i].leaf_vehicle
                for i in range(self.leaf_start, len(self.nodes))
                if self.nodes[i].keypair is not None]


if __name__ == "__main__":  # smoke test — Figure 3.11 example (N=8, isolate V3)
    lkh = PQCLogicalKeyHierarchy(level=768)
    lkh.build([1, 2, 3, 4, 5, 6, 7, 8])
    K_grp, c_root = lkh.encapsulate_group_key()
    print(f"Built N=8 tree, depth={lkh.depth}, |K_grp|={len(K_grp)}")
    print("cost before:", lkh.rekey_cost())

    tr = lkh.isolate_and_rekey(3)
    print(f"Isolated V3: refreshed {len(tr['refreshed_nodes'])} nodes, "
          f"kyber_ops={tr['kyber_ops']} (expected ceil(log2 8)={tr['expected_log2N']})")
    print("active after isolation:", lkh.active_vehicles())
    print("cost after:", lkh.rekey_cost())
